components/fxa-client/src/lib.rs

```diff
@@ -130,6 +133,45 @@ impl FxaServer {
             Self::Custom { url } => url,
         }
     }
+
+    fn from_url(url: &Url) -> Self {
+        let origin = url.origin();
+        // Note: we can call unwrap() below because parsing content_url for known servers should
+        // never fail and `test_try_from_url` tests this.
+        if origin == Url::parse(Self::Release.content_url()).unwrap().origin() {
+            Self::Release
+        } else if origin == Url::parse(Self::Stable.content_url()).unwrap().origin() {
+            Self::Stable
+        } else if origin == Url::parse(Self::Stage.content_url()).unwrap().origin() {
+            Self::Stage
+        } else if origin == Url::parse(Self::China.content_url()).unwrap().origin() {
+            Self::China
+        } else if origin == Url::parse(Self::LocalDev.content_url()).unwrap().origin() {
+            Self::LocalDev
+        } else {
+            Self::Custom {
+                url: origin.ascii_serialization(),
+            }
+        }
+    }
+}
+
```

Comment:
nit: can this be a `From<&Url>` impl instead?

Reply:
Sounds good, just updated the code